internal/app/machined/pkg/controllers/k8s/static_pod_server.go

```diff
@@ -0,0 +1,177 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package k8s
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/resource"
+	"go.uber.org/zap"
+	"gopkg.in/yaml.v3"
+
+	"github.com/talos-systems/talos/pkg/machinery/resources/k8s"
+)
+
+// StaticPodServerController renders all static pod definitions as a PodList and serves it as YAML via HTTP.
+type StaticPodServerController struct{}
+
+// Name implements controller.Controller interface.
+func (ctrl *StaticPodServerController) Name() string {
+	return "k8s.StaticPodServerController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *StaticPodServerController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: k8s.NamespaceName,
+			Type:      k8s.StaticPodType,
+			Kind:      controller.InputWeak,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *StaticPodServerController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: k8s.StaticPodServerStatusType,
+			Kind: controller.OutputExclusive,
+		},
+	}
+}
+
+type pod map[string]interface{}
+
+type podList struct {
+	Kind string `json:"kind,omitempty" protobuf:"bytes,1,opt,name=kind"`
+
+	Items []pod `json:"items" protobuf:"bytes,2,rep,name=items"`
+
+	APIVersion string `json:"apiVersion,omitempty" protobuf:"bytes,3,opt,name=apiVersion"`
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo,cyclop
+func (ctrl *StaticPodServerController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	shutdownServer, buildPodListChan, serverError, err := ctrl.createServer(ctx, r, logger)
+	if err != nil {
+		return fmt.Errorf("failed to start http server to serve static pod list: %w", err)
+	}
+
+	defer shutdownServer()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case err := <-serverError:
+			return fmt.Errorf("http server closed unexpectedly: %w", err)
+		case sendPodListChan := <-buildPodListChan:
+			staticPodList, err := buildPodList(ctx, r)
+			if err != nil {
+				logger.Error("error building static pod list", zap.Error(err))
+			}
+
+			sendPodListChan <- staticPodList
+		case <-r.EventCh():
```

Comment:
the logic here seems to be more complicated than it should be.

on reconcile event `<-r.EventCh()` we can rebuild the pod list in `[]byte` and it keep it e.g. in variable with a mutex around it; webserver can grab the mutex, get the slice and write it to the wire. (we don't need to rebuild it in other cases, as it hasn't changed).

channel ping-pong here is fine, but seems to be an overcomplication for what we need